core/src/main/scala/com/salesforce/op/evaluators/OpRegressionEvaluator.scala

```diff
@@ -86,6 +144,55 @@ private[op] class OpRegressionEvaluator
       .setMetricName(metricName.sparkEntryName)
       .evaluateOrDefault(dataUse, default = default)
   }
+
+  /**
+   * Gets the histogram of the signed percentage errors
+   *
+   * @param data Data to use
+   * @return Sequence of bin counts of the histogram
+   */
+  private def calculateSignedPercentageErrorHistogram(data: Dataset[_]): Array[Long] = {
+    // Prep data
+    val predictionsAndLabels = data
+      .select(col(getPredictionValueCol).cast(DoubleType), col(getLabelCol).cast(DoubleType))
+      .rdd
+      .map { case Row(prediction: Double, label: Double) => (prediction, label) }
+
+    // If we need to set the scaledErrorCutoff smartly, use the label data for that
+    if ($(smartCutoff)) {
+      val cutoff = calculateSmartCutoff(predictionsAndLabels)
+      log.info(s"Smart scaledErrorCutoff was determined to be: $cutoff")
+      setScaledErrorCutoff(cutoff)
+    }
+
+    val errors: RDD[Double] = predictionsAndLabels.map(x => calculateSignedPercentageError(x._1, x._2))
+    errors.histogram($(signedPercentageErrorHistogramBins))
+  }
+
+  /**
+   * Smartly calculates the scaledErrorCutoff
+   *
+   * @param predictionAndLabels  Data set containing predictions and labels
+   * @return Suggested cutoff level for scaledErrorCutoff
+   */
+  protected def calculateSmartCutoff(predictionAndLabels: RDD[(Double, Double)]): Double = {
+    val meanAbsoluteLabel = predictionAndLabels.map(_._2.abs).mean()
+    // Take the max with scaledErrorCutoff to avoid a cutoff of 0 if labels are all 0.
+    ($(smartCutoffRatio) * meanAbsoluteLabel) max $(scaledErrorCutoff)
+  }
+
+  /**
+   * Calculates the signed percentage error, with cutoff logic to avoid large results
+   * due to division by small numbers.
+   *
+   * @param prediction Predicted value
+   * @param label      Actual value
+   * @return Signed percentage error
+   */
+  private def calculateSignedPercentageError(prediction: Double, label: Double): Double = {
+    100.0 * (prediction - label) / (label.abs max $(scaledErrorCutoff))
```

Comment:
accessing spark parameters is actually quite inefficient - please access it once outside the map this calculation is called in and pass the value in as a param.

```
private def calculateSignedPercentageError(prediction: Double, label: Double, scldErrorCuttoff: Double): Double = {
    100.0 * (prediction - label) / (label.abs max scldErrorCutoff)
```

Reply:
Ooh, didn't know that, good call. 👍

Reply:
Fixed